src/linearAlgebra/Vector.h

```diff
@@ -70,24 +187,150 @@ namespace dftefe
+       * storage will point to NULL after a call to this function. Accessing the
+       * input storage pointer will lead to undefined behavior.
+       *
+       */
+      void
+      setStorage(std::unique_ptr<Storage> &storage);
+
+      /**
+       * @brief Returns a VectorAttributes object that stores various attributes
+       * (e.g., Serial or Distributed, number of components, etc)
+       *
+       * @return const reference to the VectorAttributes
+       */
+      const VectorAttributes &
+      getVectorAttributes() const;
+
+      //
+      // virtual functions
+      //
 
       /**
        * @brief Returns \f$ l_2 \f$ norm of the Vector
        * @return \f$ l_2 \f$  norm of the vector as double type
        */
-      double
-      l2Norm() const;
+      virtual double
+      l2Norm() const = 0;
 
       /**
        * @brief Returns \f$ l_{\inf} \f$ norm of the Vector
        * @return \f$ l_{\inf} \f$  norm of the vector as double type
        */
-      double
-      lInfNorm() const;
+      virtual double
+      lInfNorm() const = 0;
+
+      virtual void
+      updateGhostValues(const size_type communicationChannel = 0) = 0;
+
+      virtual void
+      accumulateAddLocallyOwned(const size_type communicationChannel = 0) = 0;
+
+      virtual void
+      updateGhostValuesBegin(const size_type communicationChannel = 0) = 0;
+
+      virtual void
+      updateGhostValuesEnd() = 0;
+
+      virtual void
+      accumulateAddLocallyOwnedBegin(
+        const size_type communicationChannel = 0) = 0;
+
+      virtual void
+      accumulateAddLocallyOwnedEnd() = 0;
+
+    protected:
+      /**
+       * @brief Constructor
+       *
+       * @param[in] storage reference to unique_ptr to Vector::Storage (i.e.,
+       * MemoryStorage) from which the Vector to transfer ownership.
+       * @param[in] vectorAttributes const reference to VectorAttributes object
+       * that contains certain properties of the Vector (e.g., serial or
+       * distributed, number of components etc.).
+       * @param[in] globalSize global size of the vector (i.e., the number of
+       * unique indices across all processors).
+       * @param[in] locallyOwnedSize size of the part of the vector for which
+       * the current processor is the sole owner (see top for explanation). For
+       * a SerialVector, the locallyOwnedSize and the the globalSize are the
+       * same.
+       * @param[in] ghostSize size of the part of the vector that is owned by
+       * the other processors but required by the current processor. For a
+       * SerialVector, the ghostSize is 0.
+       *
+       * @note Since we are passing the ownership of the input storage to the Vector, the
+       * storage will point to NULL after a call to this Constructor. Accessing
+       * the input storage pointer will lead to undefined behavior.
+       */
+      Vector(std::unique_ptr<Storage> &storage,
+             const VectorAttributes &  vectorAttributes,
```

Comment:
@bikashkanungo should we remove vectorAttributes from the constructor as it is already determined by the size parameters

Reply:
@dsambit Yes, we can remove the vectorAttributes  and use the (locallyOwnedSize==globalSize && ghostSize==0) to set it to Serial, otherwise as Distributed

Reply:
I will make that change then.